### DISCUSSION TITLE
Pure python tools

### DISCUSSION THREAD

**Discussion by gswifort** [CEXT-Dan/PyRx#67] [Ideas]
Hi, as I understand it, PyRx primarily maps C++ functions to python.
How do you feel about adding tools written in pure Python to make working with C++ functions easier?
I mean, for example, this boilerplate code:

```py
from pyrx_imp import Db

db = Db.curDb()

for id in db.objectIds(Db.Text.desc()):
    text = Db.Text(id)
    print(text.textString())
```

For example, you can create a function:

```py
def get_objects(
    container: Db.Database | Db.BlockTableRecord | Db.Group | Ed.SelectionSet,
    object_type: Db.Entity,
):
    for id in container.objectIds(object_type.desc()):
        yield object_type(id)

```

which can be used as below:

```py
for text in get_objects(db, Db.Text):
    print(text.textString())
```

Additionally, this will make it easier to debug the code, C++ exceptions are often incomprehensible at first glance, and extending the `get_objects` function with type checking etc. will make debugging much easier for beginners

**Comment by CEXT-Dan** [comment]
I think it’s great! It’s only natural to want to extend wrappers with syntax that’s more fitting to Python. 

My original thoughts were to use similar syntax as ObjectARX, easy to port projects, use the documentation.
I.e. for additional help on Ed.Core.coordFromPixelToWorld, just copy&paste *coordFromPixelToWorld in Autodesk help…
https://help.autodesk.com/view/OARX/2024/ENU/?query=*coordFromPixelToWorld

Regarding exceptions, they currently include the things I need for support. Error code, Line number, function, and file. From here, I know exactly where to debug from in C++. 

But yeah, I would fully expect Python users to want to add cool layers of abstraction on top. It’s better not done by me, as I’ve already been accused of having a C++ bias lol. 

I haven’t thought much about library contributions, I guess I would have to make some guidelines. I.e. Unit tests and documentation

**Reply by gswifort** [reply]
That's right, copying ARX functions 1:1 is the best solution. Python tools must be clearly separated from ARX modules, e.g.:
```
PyRx/
├── core/
│   ├── PyRx
│   ├── PyDb
│   ├── PyAp
│   └── ...
└── tools/
    └── ...
```

We are waiting for the guidelines! 🚀

**Reply by CEXT-Dan** [reply]
How do you imagine pure python tools loading?

Currently, everything that’s not C++, and can be imported, is in PyRxStubs. I.e. ActiveX wrappers and PyRxDebug
I plan on adding more since SheetSets and ETransmit are also ActiveX

I would rather not have another environment variable besides the stubs path.  
Though, stubs is not a great name. It could be renamed something like resources

How do you imagine the imports looking? Users should be able to choose, or mix and match what to use.

Note, the folder structure is quite difficult to change, I.e. I can rename folders like PyRx to PyRxCore, but don’t want to move the levels.
The C++ projects use envs like $(SolutionDIR), so PyRx.sln must remain top

**Reply by gswifort** [reply]
I think the best solution would be to change the approach to installation.
The project can be hosted on [PyPI](https://pypi.org/) so that it can be installed using `pip install`.
After installing the package, we would have access to modules written in Python, the installation could be global or in a virtual environment.
The `Bin` folder would be downloaded together with the python modules, and after executing the command, e.g. `python -m pyrx bundle install`, the `Bin` directory would be copied to `%localappdata%/...` (i.e. it would do the same thing as the ` .msi` installer does now ). A similar approach is used in the [`xlwings`](https://docs.xlwings.org/en/stable/index.html) project, where using [`xlwings addin install`](https://docs.xlwings.org/en/stable/installation.html#xlwings-excel-add-in) the excel addin is installed.
We would further use what we obtained in #1, if Cad was launched from the virtual environment, it would have access to the `pyrx` library (pure python) from the virtual environment, and if not, the packages found on the basis of `PyRx.ini` would be used.
In the module loaded with `PYLOAD` we would then have:

```py
from pyrx.tools import get_objects
```

ActiveX modules can also be moved:

```py
from pyrx import AxApp25 as Ax
```

I think eventually stubs should also be included in the pyrx library, we need to refer to [PEP 561](https://peps.python.org/pep-0561/)

**Reply by gswifort** [reply]
Of course PyPI is not required, you can also install it with `pip install git+https://github.com/CEXT-Dan/PyRx.git`

**Reply by CEXT-Dan** [reply]
Sounds like an awesome plan!
This would automatically download wxPython as a requirement, right?
I’ll start working in this direction.
Thanks for the hints!

**Reply by gswifort** [reply]
Yes, dependencies will be included.
By the way - why do you require `python 3.12`? It would be nice to be able to use older versions, especially since `python>3.8` requires at least Windows 10.
The examples use the `match...case` construct, which was introduced in `python 3.12`, but is that the only reason?

**Reply by gswifort** [reply]
I suggest creating a new branch for this package. It should be developed to some extent before being published in an official release.

**Reply by CEXT-Dan** [reply]
It’s a C++ dependency issue, I have a hard link with python312.lib and I use boost::python wrappers that are also linked with python312.lib

https://github.com/CEXT-Dan/PyRx/blob/4bafe887baff7cb5b275017480e73557364d25cb/PyRxCore/StdAfx.h#L137-L143

At the time I chose the latest because I knew I would be stuck with this version for a while. It’s theoretically possible to have a CMAKE system that would build, but that would be a huge task, and it would take an hour to build the whole package. I considered using Pybind11 which is faster to build, but 5x slower in performance. someday I’ll revisit these, maybe check out NanoBind. 

We are really fast; I don’t know if you’ve seen this thread. But I know the .NET guys were sweating being crushed by python
https://www.theswamp.org/index.php?topic=59487.msg620757#msg620757

I think choosing boost::python was the right decision, it has bindings with NumPy, so I can access this library outside the interpreter... someday

Anyway, I think I would probably need a build/test server to support multiple python versions X multiple cad versions

**Reply by gswifort** [reply]
The minimum structure should probably look like this:

```
root/
├── pyrx/ # python modules
│ ├── __init__.py # import PyDb, PyEd etc.
│ └── ... # other packages
└── pyrx-stubs/
 ├── PyDb.pyi # or Db.pyi if `import PyDb as Db`
 ├── PyEd.pyi
 └── ...
setup.py
```

in the `pyrx/__init__.py` we will import C++ modules so everything will be correct

```py
# pyrx/__init__.py

# someone may want to test python modules outside of cad
# then these imports will not work, you should check
# whether we are in the cada process or not and if not, skip the ImportError
import PyAp as Ap
import PyDb as Db # in this case we need to change PyDb.pyi to Db.pyi
import PyEd as Ed
import PyGe as Ge
import PyGi as Gi
import PyGs as Gs
import PyPl as Pl

import PyRx as Rx

# this module is to hide Pylance 'module not found' warnings
# Pylance cant read Python embedded in C++


import importlib.util
if importlib.util.find_spec("PyBrxCv") is not None:
    import PyBrxCv as Cv

import importlib
flag = importlib.import_module('win32com')
if flag is not None:
    try:
        host = Ap.Application.hostAPI()
        if host == "BRX24":
            import BxApp24 as Ax
        elif host == "GRX24":
            import GxApp24 as Ax
        elif host == "ZRX24":
            import ZxApp24 as Ax
        elif host == "ZRX25":
            import ZxApp25 as Ax
        elif host == "ARX24":
            import AxApp24 as Ax
        elif host == "ARX25":
            import AxApp25 as Ax
    except ImportError:
        pass
        
```

```py
# setup.py

from setuptools import find_packages, setup

setup(
    name="pyrx",
    version="0.1.0",
    description="...",
    packages=find_packages(include=["pyrx", "pyrx-stubs"]),
)
```

Sample PYLOADed module:

```py
from pyrx import Db
# or
from pyrx import PyDb as Db
...
```

**Reply by CEXT-Dan** [reply]
Thinking about this more. If I did this `pip install git+github.com/CEXT-Dan/PyRx.git`
Then I would have to have the ARX binaries committed, and the package would be created from main. this doesn't seem ideal 

PEP 561, it seems no one is following this? I only found one package that did. wxPthon, SciPy, NumPy all have .py and .pyi in the same folders, and rightly so. Separating them out seems like a maintenance hassle.

I’m thinking of having some sort of packaging script that’s run in visual studio, I would replace the msi with this package, and or post it on pypi

**Reply by gswifort** [reply]
In terms of installation from github, I wanted to install from a specific git tag from the `.whl` file for a specific release (just as now you add a `zip` with `Bin` to the release, but of course the best solution is to publish on PyPI.


You're right, a separate stubs package is unnecessary.
I think this is what a `pyrx` package should look like, regardless of the distribution method:

```
pyrx/
  __init__.py  # import PyDb, PyEd etc.
  PyDb.pyi
  PyEd.pyi
  ...  # other pure python packages
```

```py
# __init__.py

import PyRx as Rx
import PyGe as Ge
import PyGi as Gi
import PyDb as Db
import PyAp as Ap
import PyEd as Ed
import PyPl as Pl
import PyGs as Gs

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from . import PyRx as Rx
    from . import PyGe as Ge
    from . import PyGi as Gi
    from . import PyDb as Db
    from . import PyAp as Ap
    from . import PyEd as Ed
    from . import PyPl as Pl
    from . import PyGs as Gs

from .utils import get_objects  # python modules

def pure_python_func(db: Db.Database) -> Db.Text:
    ...
```

**Reply by gswifort** [reply]
@CEXT-Dan Have you already started working on installers?  If not, maybe I would try to work on this topic.

**Reply by CEXT-Dan** [reply]
That would be great, and thank you!  I really like the idea of installing from a GitHub tag, this would be a first step before publishing to PyPi. It would give me time to work on more unit tests. In fact, there are still wrappers missing.

I think the only requirement that I would have for pure Python libraries is that they have some sort of test coverage. It can/should be a separate test package for now to avoid git conflicts. But should be something I can run in debug mode to detect memory leaks

**Reply by gswifort** [reply]
I'm thinking out loud, maybe the C++ based package should be separated from the python package.  i.e. `pyrx` would work closely with `pyrx-core`, but `pyrx-core` could run independently.  The main advantage would probably be the ability for an older version of `pyrx` to work with a newer version of `pyrx-core` or vice versa.

**Reply by CEXT-Dan** [reply]
On the surface that seems fine, but when I think about it, it could quickly become a support nightmare

**Reply by gswifort** [reply]
Maybe in order not to close ourselves off (the need to change imports), we should adopt this structure for now:
```raw
repo/
├── pyrx/
│   └── ... # python modules
└── pyrx-core/
    ├── Bin/
    │   └── ...
    ├── Stubs/
    │   └── ...
    └── activex/
        └── ...
```
For now, it would be two packages, but distributed together, later we could think about separating them.

**Reply by CEXT-Dan** [reply]
I don’t understand

**Reply by gswifort** [reply]
In your repository, in addition to the current directories and files, there would be two directories:
- `pyrx`
- `pyrx_core`
The pyrx_core package would contain everything that I currently put [here](https://github.com/gswifort/PyRx/tree/73-cad-bundle-installer)
i.e. the `Bin` directory with ARX modules, the `Stubs` directory with stub files for ARX and the `activex` directory with `.py` and `.pyi` files for ActiveX.

The `pyrx` package would contain all the python tools.
An example module in `pyrx` would look like this:

```py
from pyrx_core import Db

def do_anything_with_db(db: Db.Database):
    ...
```

Currently, both directories would be in the same repository and distributed in PyPI under the name `pyrx`, but when the user installs `pip install pyrx`, two packages will be installed in the `site-packages` directory - `pyrx` and `pyrx_core`. Just like when installing `pywin32` - `win32`, `win32com`, `win32comext` etc. are installed in `site-packages`.
Later, when `pyrx_core` is already well developed and the only changes will be changes resulting from new versions of Autocad etc. and bugs - it will be possible to separate these two packages into two repositories, but `pyrx` would have `pyrx_core` as a dependency.

What's the point of all this?
I imagine the Python package `pyrx` will be in development for much longer than `pyrx_core`.
If someone creates a project based on `pyrx 1.1.0`, they may not be ready for the changes introduced in newer versions, and at the same time they may want to update their project to work with the new version of Autocad.
If both of these packages were distributed under one version, someone who wanted to run their project on a new version of Autocad would have to take into account all the changes occurring in the Python packages.

In my opinion, as long as everything is in one repository, it doesn't matter for maintenance whether it's two packages or one.

**Reply by CEXT-Dan** [reply]
No, I don’t want separate packages. It should be tested, packaged, and supported as a single version.

**Comment by CEXT-Dan** [comment]
ARX and ActiveX

I you look at the API in the aggregate, it contains both native and ActiveX,
SheetSets, ETransmit, Properties, Standards Manager, DesignCenter etc.
Wrapping ActiveX in C++ is not fun at all, while generators for the most part are pretty good, so I will leverage those when I can.
There may be a future where the ARX side calls a Python wrapper and vice versa 
Point is, PyRx-Core API is combination of both ARX and ActiveX.

Just a thought, but pyrx-core/ could just include everything that’s core related. 
Or have pyrx-core/Bin and pyrx-core/stubs
